ext/drreg/drreg.c

```diff
@@ -441,6 +720,65 @@ drreg_event_bb_insert_late(void *drcontext, void *tag, instrlist_t *bb, instr_t
     }
 
     /* Before each app read, or at end of bb, restore spilled registers to app values: */
+#ifdef X86
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++) {
+        restored_for_simd_read[SIMD_IDX(reg)] = false;
+        if (!pt->simd_reg[SIMD_IDX(reg)].native) {
+            if (drmgr_is_last_instr(drcontext, inst) ||
+                instr_reads_from_reg(inst, reg, DR_QUERY_INCLUDE_ALL) ||
+                /* i#1954: for complex bbs we must restore before the next app instr */
+                (!pt->simd_reg[SIMD_IDX(reg)].in_use &&
+                 ((pt->bb_has_internal_flow &&
+                   !TEST(DRREG_IGNORE_CONTROL_FLOW, pt->bb_props)) ||
+                  TEST(DRREG_CONTAINS_SPANNING_CONTROL_FLOW, pt->bb_props)))) {
+                if (!pt->simd_reg[SIMD_IDX(reg)].in_use) {
+                    res = drreg_restore_reg_now(drcontext, bb, inst, pt, reg);
```

Comment:
Same problem again here. Because we're iterating over the applicable range that resembles zmm, we will end up restoring zmm, even if we had spilled only xmm.

Reply:
Let me see.

Reply:
Doesn't drreg_restore_reg_now fetch the reg spilled:

Line 1881: `reg_id_t spilled_reg = pt->simd_slot_use[pt->simd_reg[SIMD_IDX(reg)].slot];`

Reply:
Ok. Still very paranoid about the xmm/ymm/zmm mix, in particular because of evex masks and _all_ writes to ymm writing null to zmmh, and therefore still tending to think if it was better to only support xmms. But you're right here, I think it's ok, thanks for resolving my concern.

Reply:
I agree with you that this is a concern 🤔 So the idea of the overall PR was to support XMM registers, but also include sufficient boilerplate code to support other registers to try and minimise future changes and force us to implement a modular design. My initial code I use for my work was based on XMM only - I modified this based on the discussion on the issues page, and therefore you are right to say that the current code has not been scrutinised enough for YMM/ZMM.

I need to check this, but all instructions that deal solely with XMM registers are not EVEX right? A ZMM register is cleared only on YMM writing. So perhaps we need to have two main mixes instead of three in regards to LIVE tracking, namely: 1) XMM liveliness 2) Y/ZMM liveliness. In other words, any YMM reservation defaults to ZMM reservation but to keep spilling under control, we also perform efficient spills for XMM using movdqa. 

We need to discuss the following as I am not sure, but we can focus on XMM for now, with the existing boiler-plate code, and mark the design of other SIMD registers as experimental for now. Besides, there are asserts in place to stop the user from reserving YMM and ZMM anyways. I think this is the quickest route to getting the code in.

Reply:
> all instructions that deal solely with XMM registers are not EVEX right? A ZMM register is cleared only on YMM writing

No, evex writes to xmm will null zmmh and ymmh. Vex writes will null ymmh.

Reply:
Aha, so if we do indeed consider the two mixes/instead of XMM, YMM and ZMM separately, it will help XMM users who solely consider SSE instructions hence the movdqa. You highlight a good point and this need to be documented. 

I guess the decision depends on your use case. How do you intended to use the reserved XMM register. Will your future instrumented code end up clobbering the upper bytes in YMM and ZMM?

Reply:
> I guess the decision depends on your use case. How do you intend to use the reserved XMM register? Will your future instrumented code end up clobbering the upper bytes in YMM and ZMM?

This is separate. If the user reserves xmm, the user is responsible to use it only as xmm. My concerns are limited to drreg managing the state correctly.

Reply:
Aha, I agree it is a separate issue also. The reason I ask is bc if you are only going to use SSE instructions in your use-case, without corrupting ymmh and zmmh due to evex/vex instructions, then we can consider supporting XMM only for now. If you are going to use evex/vex instructions, then we will need to be able to support the other SIMD registers.

Reply:
I think with SSE and xmm only, it is ok because SSE preserves ymmh and zmmh, and there are no partial writes to part of xmm's, unless the destination is a xmm.